### PR TITLE
V2 write issue (implements proposal in issue #554)

### PR DESCRIPTION
**Pull request by christianknoepfle**
This PR implements the proposal in issue #554 and resolves issue #547 and issue #539. See the proposal for technical approach.
In addition the README.md was enhanced to reflect the changes

### CODE REVIEW

**Comment by nightscape** [comment]
Hey @christianknoepfle, thanks for the PR!
I just triggered the Github build.
I'm not too familiar with the v2 implementation, so @quanghgx would have to chime in with a code review here.

**Comment by christianknoepfle** [comment]
The test failures on the spark 3.x are related to reading a partitioned excel file structure and that leads to a crash deep in poi. Unsure how to force this behavior on every test run so it is easier to debug.  Any ideas?

**Comment by christianknoepfle** [comment]
Update: Added a test to get the java.util.ConcurrentModificationException when executing locally by reading a folder with100+ partitions. So I can investigate further.

**Comment by christianknoepfle** [comment]
I found the issue of the java.util.ConcurrentModificationException and added a fix for it.

See here : src/main/scala/com/crealytics/spark/v2/excel/ExcelHelper.scala

The comment block for ExcelHelper.getWorkbook() . The issue was present prior the changes of this PR but didn't show up until now (it occurs rarely and random, but with more partition testcases chances increased). 

Note that there is also an alternative implementation mentioned in the comment block. Since I am not a poi expert, I have no idea if this alternative implementation would be sufficient (I didn't get why the providers are re-registered anyhow, because those factory classes do not have any state)

Could you pls run the workflow so we see if the issue is gone in your pipeline too (it worked on my end, but last time that was the case too)

File: src/main/scala/com/crealytics/spark/v2/excel/ExcelHelper.scala [hunk 1]
```diff
@@ -200,15 +200,43 @@ class ExcelHelper(options: ExcelOptions) {
...
+        Singleton.INSTANCE.provider at the same time (i.e. while it is changing).
+
+        This can lead to a java.util.ConcurrentModificationException exception while looping
+        over the mentioned data structure  (code fragment
+        at org.apache.poi.ss.usermodel.WorkbookFactory.wp(WorkbookFactory.java:327)
+
+        In case we need to exchange the provider on each call, we need to move the
+        synchronized so it guards the call to the .wp() function too.
+     */
+    if (!configurationNeeded.getAndSet(true)) {
```
Comment (nightscape) [inline src/main/scala/com/crealytics/spark/v2/excel/ExcelHelper.scala:257]: Nice!

**Comment by christianknoepfle** [comment]
after discussing registerProviders() with @cristichircu here (https://github.com/crealytics/spark-excel/issues/480#issuecomment-1065868049) I enabled the code to do registration only once (also moved the call to apply() which seemed to me a better place).  The CI is all green on my end. How can we move forward with this PR?

File: src/test/scala/com/crealytics/spark/v2/excel/LocalFileTestingUtilities.scala [hunk 2]
```diff
@@ -55,11 +54,10 @@ trait LocalFileTestingUtilities {
 
     fixture
   }
-
-  def createTemporaryDirectory(name: String): File = {
+  def createTemporaryDirectory(prefix: String): File = {
```
Comment (nightscape) [inline src/test/scala/com/crealytics/spark/v2/excel/LocalFileTestingUtilities.scala:57]: Could we use the built-in mechanism for creating temp directories?
https://docs.oracle.com/javase/7/docs/api/java/nio/file/Files.html#createTempDirectory%28java.nio.file.Path,%20java.lang.String,%20java.nio.file.attribute.FileAttribute...%29

Comment (christianknoepfle) [reply src/test/scala/com/crealytics/spark/v2/excel/LocalFileTestingUtilities.scala]: done

File: src/test/scala/com/crealytics/spark/v2/excel/DataFrameWriterApiComplianceSuite.scala [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+
+        dfCsv.write
+          .format("excel")
+          .option("path", targetDir)
+          .option("header", value = true)
+          .mode(writeMode)
+          .save()
+
+        val listOfFiles = getListOfFilesFilteredByExtension(targetDir, "xlsx")
+        assert(listOfFiles.nonEmpty, s"Expected a single file but got more: $listOfFiles")
```
Comment (nightscape) [inline src/test/scala/com/crealytics/spark/v2/excel/DataFrameWriterApiComplianceSuite.scala:64]: The assertion logic and message don't seem to align.

**Comment by nightscape** [comment]
@quanghgx would you have time to do a review here?

Afaiu it not possible to write a single `.xlsx` file with a specified name, but instead it will always be the default Spark folder structure, right?
So users who want a single file need to do a `.repartition(1)` to get a single file and then move it out of the subdirectory and rename it, or use the v1 writer.
I would be ok with this, as long as users still can write to and read from a single file (which is what I mostly use it for).
In the mid-term, it would be nice to provide some helper function that allows users to read/write single files with the v2 API, so that we can get rid of v1.

**Comment by cristichircu** [comment]
Hi @nightscape! I started a discussion about this sort of functionality a while back: https://github.com/crealytics/spark-excel/discussions/549. Do you think this would be useful to others? Can you please share your thoughts on the proposed approach?

**Comment by quanghgx** [comment]
Hi Martin and @christianknoepfle 
 
Sorry for the long delay from my side. I was a bit tied up in the last couple of months.
 
About single output file (@nightscape): For spark data source 3.x API, if we reuse the same base classes with built in data sources (like CSV, JSON, Parquet e.g), it is not easy to write to a single excel file. However, in concept, these base classes are not the only route. I did have a couple of other spark data sources (for work), those do not depend on spark built-in base classes, so we can write to any structure we need.
 
Supporting multiple spark versions, while reuse spark built-in supportive classes, is quite a burden. To be fair, these built-in classes seem not intended to be reused by external data sources like spark-excel. I do think that we may need to move away from built-in supportive classes and also reduce the number of supported spark versions. Spark-Excel feature set seems stable by now, so if users need spark-excel with older spark-version, they may need to pick an appropriate (older) spark-excel version.
So, @christianknoepfle, 1+ from me if you need to cleanup some fragmented folders.

Sincerely,

File: build.sbt [hunk 4]
```diff
@@ -6,7 +6,7 @@ crossScalaVersions := Seq("2.12.15", "2.13.8")
 
 scalaVersion := crossScalaVersions.value.head
 
-lazy val sparkVersion = "3.2.1"
+lazy val sparkVersion = "3.0.1"
```
Comment (quanghgx) [inline build.sbt:9]: Maybe, keep the latest, stable spark version as the default?

**Comment by quanghgx** [comment]
Given that remain comments are minor.
1+ with merging, @christianknoepfle and @nightscape

**Comment by nightscape** [comment]
Ok, squash-merging it now 👍

**Pull request merged**